### PR TITLE
Add Debugger decorator to add runtime, memory used, and context to functions.

### PR DESCRIPTION
**Pull request by amshamah419**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
Added a debugger decorator to log function name, arguments, keyword arguments, runtime, memory usage, and exceptions to XSOAR/XSIAM server logs for enhanced debugging and monitoring.

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 1]
```diff
@@ -41,7 +42,7 @@ def __line__():
 
 # 43 - The line offset from the beginning of the file.
```
Comment (dorschw) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:43]:
```suggestion
# 45 - The line offset from the beginning of the file.
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 2]
```diff
@@ -12051,6 +12052,62 @@ def get_server_config():
     server_config = body.get('sysConf', {})
     return server_config
 
+
+# Type variable to indicate that the decorated function can be any callable
```
Comment (dorschw) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12056]: won't just using `Callable` work? 🤔

**Review by dorschw — COMMENTED** [review]
NICE

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 3]
```diff
@@ -12051,6 +12053,104 @@ def get_server_config():
...
+    def decorator(func: FuncType) -> FuncType:
+        @wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            # Log the timestamp and function details
+            function_name = func.__name__
+            masked_args, masked_kwargs = mask_secrets(args, kwargs, func, secret_keys)
+
+            demisto.debug(f"Calling function: {function_name}")
+            demisto.debug(f"Arguments: {masked_args}")
+            demisto.debug(f"Keyword arguments: {masked_kwargs}")
```
Comment (tkatzir) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12119]: Why not do just one `debug` call?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 4]
```diff
@@ -12051,6 +12053,104 @@ def get_server_config():
...
+            end_time = time.time()
+            runtime = end_time - start_time
+
+            # Capture end memory usage
+            end_memory = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
+            memory_used = end_memory - start_memory
+
+            # Log runtime and memory usage
+            demisto.debug(f"Function {function_name} completed in {runtime:.4f} seconds")
+            demisto.debug(f"Memory used by {function_name}: {memory_used / 1024:.2f} KB")
```
Comment (tkatzir) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12143]: Why not do just one `debug` call?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 5]
```diff
@@ -12051,6 +12053,104 @@ def get_server_config():
...
+            # Log the timestamp and function details
+            function_name = func.__name__
+            masked_args, masked_kwargs = mask_secrets(args, kwargs, func, secret_keys)
+
+            demisto.debug(f"Calling function: {function_name}")
+            demisto.debug(f"Arguments: {masked_args}")
+            demisto.debug(f"Keyword arguments: {masked_kwargs}")
+
+            # Capture the start time and memory usage
+            start_time = time.time()
```
Comment (tkatzir) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12122]: This should be move to be right before calling the function.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 6]
```diff
@@ -12051,6 +12053,104 @@ def get_server_config():
...
+            # Capture end memory usage
+            end_memory = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
+            memory_used = end_memory - start_memory
+
+            # Log runtime and memory usage
+            demisto.debug(f"Function {function_name} completed in {runtime:.4f} seconds")
+            demisto.debug(f"Memory used by {function_name}: {memory_used / 1024:.2f} KB")
+
+            # Log the result
+            demisto.debug(f"Function {function_name} result: {result}")
```
Comment (tkatzir) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12146]: Maybe do it in a `finally` block?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 7]
```diff
@@ -12051,6 +12053,104 @@ def get_server_config():
     server_config = body.get('sysConf', {})
     return server_config
 
+
+def mask_secrets(args: tuple, kwargs: dict, func: Callable, secret_keys: Optional[List[str]] = None) -> tuple:
```
Comment (tkatzir) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12057]: You can use default for `secret_keys`:
- sensitive_params, in IntegrationLogger.__init__
- keywords_to_censor, in censor_request_logs

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 2 above)
Comment (amshamah419) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12094]: Techincally yes, but doing so this way will preserve a few important things, like the original typing of the function it wraps, and is more type safe.

**Pull request closed without merging**